addons/sale/models/sale.py

```diff
@@ -631,28 +631,62 @@ def _create_invoices(self, grouped=False, final=False, date=None):
 
         # 1) Create invoices.
         invoice_vals_list = []
+        invoice_item_sequence = 0
         for order in self:
-            pending_section = None
+            prepared_pending_section = None
+            down_payments = []
 
             # Invoice values.
             invoice_vals = order._prepare_invoice()
 
             # Invoice line values (keep only necessary sections).
+            invoice_line_ids = []
             for line in order.order_line:
                 if line.display_type == 'line_section':
-                    pending_section = line
+                    prepared_pending_section = line._prepare_invoice_line()
                     continue
-                if float_is_zero(line.qty_to_invoice, precision_digits=precision):
+                if line.display_type != 'line_note' and float_is_zero(line.qty_to_invoice, precision_digits=precision):
                     continue
-                if line.qty_to_invoice > 0 or (line.qty_to_invoice < 0 and final):
-                    if pending_section:
-                        invoice_vals['invoice_line_ids'].append((0, 0, pending_section._prepare_invoice_line()))
-                        pending_section = None
-                    invoice_vals['invoice_line_ids'].append((0, 0, line._prepare_invoice_line()))
-
-            if not invoice_vals['invoice_line_ids']:
+                if line.qty_to_invoice > 0 or (line.qty_to_invoice < 0 and final) or line.display_type == 'line_note':
+                    if line.is_downpayment:
+                        down_payments.append(line)
+                        continue
+                    if prepared_pending_section:
+                        invoice_item_sequence += 1
+                        prepared_pending_section['sequence'] = invoice_item_sequence
+                        invoice_line_ids.append(prepared_pending_section)
+                        prepared_pending_section = None
+                    invoice_item_sequence += 1
+                    prepared_line = line._prepare_invoice_line()
+                    prepared_line['sequence'] = invoice_item_sequence
+                    invoice_line_ids.append(prepared_line)
+
+            # If down payments are present in SO, group them under common section
+            if len(down_payments):
+                invoice_item_sequence += 1
+                down_payments_section = {
```

Comment:
Maybe define a small tool method _prepare_downpayment_line ?

Reply:
decorated with @api.model and added to SaleOrder class